Add tests to pivot and spin calibration and fix a precision bug

theartful opened the pull request:
I'm working on a project that uses IGSIO, and I wrote some tests to make sure that calibration works as expected. Thought I would share them with you as well.

Both tests randomly generate transformations according to a "ground truth".

For pivot calibration, the ground truth is the tip position in world coordinates (should be fixed), and the position of the marker in marker coordinates (which will be mapped to -Z direction as stated in the docs). The tests work as follows:
1. Generate marker transforms by:
    * Applying a small random noise to the pivot to mimic non steady surface / hand.
    * Sample a random point on the sphere whose center is the perturbed pivot, and its radius is the stylus length. This point is the marker position.
    * Create a transform according the new pivot and marker positions.
2. Feed these transforms to the algo
3. Make sure that the computed pivot position is as close as possible to the ground truth pivot position.

For spin calibration, the ground truth is the tip position in world coordinates (fixed), the spin axis (which should be calculated), and the distance between the marker and the spin axis. The tests work as follows:
1. Generate marker transforms by:
    * Generating an initial random marker transform
    * Make two revolutions around the spin axis starting from the initial marker transform
    * Add random rotation perturbation to each transform around an axis different than the spin axis
2. Feed these transforms to the algo
3. Make sure that the angle between the computed spin axis, and the ground truth spin axis is as small as possible.

theartful committed:
Add tests to pivot and spin calibration and fix a precision bug

theartful committed:
Remove formatting changes in vtkIGSIOSpinCalibrationAlgo

theartful committed:
Change hardcoded maxIterations

lassoan commented on IGSIOCalibration/Testing/Utils.h:
```diff
@@ -0,0 +1,93 @@
+  transform->GetMatrix(rotation);
+
+  return rotation;
+}
+
+// Applies a small random rotation to a matrix to simulate tracking noise
```
Why noise is only added to the rotation, not to translation?

theartful replied:
This method is only used in spin calibration. The solver actually doesn't care about translation at all, so I opted not to add a parameter for translation noise in the test.

Regardless, this is the new signature
```C++
static vtkNew<vtkMatrix4x4> ApplyRandomPerturbation(vtkMatrix4x4* matrix, double maxAngle, double translationSigma, std::mt19937& rng)
```
and both spin and pivot calibration use it to add noise now. For pivot calibration, `maxAngle` is zero (as the rotation is random to begin with). For spin calibration `translationSigma` is zero (as the solver doesn't depend on translation).

theartful replied:
I realized this later, but I think neither `markerRadius` nor `stylusLength` affect the spin calibration solver as well. They help with imitating real marker transformations though. So maybe it's a good idea to add translation noise?

lassoan commented on IGSIOCalibration/Testing/vtkIGSIOPivotCalibrationTest.cxx:
```diff
@@ -0,0 +1,263 @@
+  double StylusLength;           // Length from tracker origin to tip
+};
+
+static vtkVector3d GenerateRandomDirection(std::mt19937& rng)
```
Shouldn't we use the implementation in Utils.h?

theartful replied:
Yep, I refactored common methods to Utils.h then forgot to use it in the pivot calibration test. Done.

lassoan commented on IGSIOCalibration/Testing/vtkIGSIOPivotCalibrationTest.cxx:
```diff
@@ -0,0 +1,263 @@
+
+struct GroundTruth
+{
+  vtkVector3d TipPositionWorld;  // Fixed pivot point in world coordinates
```
```suggestion
  vtkVector3d TipPosition_World;  // Fixed pivot point in world coordinates
```

lassoan commented on IGSIOCalibration/Testing/vtkIGSIOPivotCalibrationTest.cxx:
```diff
@@ -0,0 +1,263 @@
+struct GroundTruth
+{
+  vtkVector3d TipPositionWorld;  // Fixed pivot point in world coordinates
+  vtkVector3d TipPositionMarker; // Fixed pivot point in marker coordinates
```
```suggestion
  vtkVector3d TipPosition_Marker; // Fixed pivot point in marker coordinates
```

lassoan commented on IGSIOCalibration/Testing/vtkIGSIOPivotCalibrationTest.cxx:
```diff
@@ -0,0 +1,263 @@
+{
+  vtkVector3d TipPositionWorld;  // Fixed pivot point in world coordinates
+  vtkVector3d TipPositionMarker; // Fixed pivot point in marker coordinates
+  vtkVector3d TipDir;            // The unit vector direction from the stylus marker to the tip in marker coordinates
```
```suggestion
  vtkVector3d TipDir_Marker;            // The unit vector direction from the stylus marker to the tip in marker coordinates
```

lassoan commented on IGSIOCalibration/vtkIGSIOSpinCalibrationAlgo.cxx:
```diff
@@ -176,8 +176,11 @@ igsioStatus vtkIGSIOSpinCalibrationAlgo::DoSpinCalibrationInternal(const std::ve
+  // "A" is a positive semidefinite matrix by construction (as it's the summation of positive semidefinite matrices in the form R^t * R)
+  // This means that all eigenvalues have to be >= 0. But due to precision errors in vnl_symmetric_eigensystem_compute, a negative eigenvalue
+  // might be computed that is very close to zero. So we just approximate them to zero.
+  error = sqrt(std::max(eigenvalues(0), 0.0) / markerToTransformMatrixArray.size());
```
It could make sense to specify a tolerance. If absolute value of the eigenvalue is smaller than the tolerance then we could use the tolerance value instead of the negative value (returning exactly 0 error when we know that there is numerical inaccuracy would be somewhat misleading). If the eigenvalue<-tolerance then it could make sense to log a warning (as it should never happen).

theartful replied:
This is how it looks now
```C++
  const double tolerance = 1e-9;
  const double minEigenvalue = eigenvalues(0);
  if (minEigenvalue < -tolerance) {
    LOG_WARNING("Expected the minimum eigenvalue of the aggregate instantaneous rotation matrices to be positive. Instead got: " << minEigenvalue);
  }

  //set the RMSE
  error = sqrt(std::max(std::abs(minEigenvalue), tolerance) / markerToTransformMatrixArray.size());

```

lassoan reviewed (commented):
Very nice work, thank you. I've just added a few nitpicky comments.

theartful committed:
Address comments

1. Use _World and _Marker suffix
2. Better handle the case of negative eigenvalues (which happen in tests)
3. Unify how noise is applied

theartful committed:
Use normal distribution for translation noise

theartful committed:
minEigenvalue instead of eigenvalues(0)

theartful committed:
Use _World and _Marker suffix for other variables as well

theartful committed:
Put the marker at stylusLength distance on the spin axis

The old calculation of "h" is unjustified

theartful committed:
Revert formatting changes and simplify tip position calculation

The old calculation via Pythagoreus is unjustified
